plot/plot.py

```diff
@@ -20,8 +32,73 @@ def window_hanning(window):
     '''
     return np.hanning(len(window))*window
 
-def psd(samples, nfft=None, sample_rate=None, window=window_hanning, noverlap=None,
-        detrend_func=None, pad_to=None, scale_by_freq=None, sides=None):
+def next_power_of_2(val):
+    """
+        Return the next integer that is a power of two
+
+        Params
+        ------
+        val : int
+    """
+
+    return int(2**(int(np.log2(val)) + 1))
+
+def is_power_of_2(val):
+    """
+        Checks wheters the value is a power of two
+    """
+    return (val != 0) and ((val & (val -1)) == 0)
+
+
+def psd(samples, sampling_frequency, center_frequency, nfft=None):
+    """
+        Plot the power spectral density.
+    """
+    ## Define Parameters
+    # Length of signal
+    nsamples = len(samples)
+
+    # Number of FFT points
+    if nfft is None:
+        nfft = next_power_of_2(nsamples)
+
+    if nfft is not is_power_of_2(nfft):
+        nfft = next_power_of_2(nfft)
+
+    if np.ndim(samples) > 1:
+        samples = samples.reshape((samples.shape[0]*samples.shape[1],))
+
+    # Frequency plotting vector
+    freqs = sampling_frequency/2*(np.arange(-1, 1-2/nfft, 2/nfft, dtype=type(center_frequency)))
+    freqs = np.asarray(freqs) + center_frequency
+    ## Analysis
+
+    # analyze spectrum
+    indeces = np.arange(nfft/2-1, nfft-1, dtype=int)
+    if np.ndim(samples) == 1:
+        power = fourier.fftshift(fourier.fft_vectorized(samples, nfft))
+        power = fourier.fftshift(power)
+    elif np.ndim(samples) == 2:
+        samples = samples.reshape((samples.shape[0]*samples.shape[1],))
+        power = np.asarray(fourier.fft_vectorized(samples))
+        # power = fourier.fftshift(power)
+    else:
+        raise ValueError("Only 1 and 2 dimensional sample arrays are supported right now.")
+
+    power = np.abs(power)/np.sqrt(nsamples*sampling_frequency)
+
+
+    # Convert to dBm/Mz
+    power = 10*np.log10(power/center_frequency)
+
+    freqs = [freqs[i] for i in indeces]
+    power = [power[i] for i in indeces]
+
+    return freqs, power
+
+
+def opsd(samples, nfft=None, sample_rate=None, window=window_hanning, noverlap=None,
+         detrend_func=None, pad_to=None, scale_by_freq=None, sides=None):
     """
         Plot the power spectral density.
     """
```

Comment:
Op regel 210 wordt pylint issue R1705 gedisabled, echter is de issue terecht.
Volgens mij zijn er nog wel manieren om de 'detrend'-functie te verbeteren.